File: Samples/Menu3D/C++/Src/MultiplierConverter.cpp
////////////////////////////////////////////////////////////////////////////////////////////////////
// NoesisGUI - http://www.noesisengine.com
// Copyright (c) 2013 Noesis Technologies S.L. All Rights Reserved.
////////////////////////////////////////////////////////////////////////////////////////////////////


#include "MultiplierConverter.h"

#include <NsCore/ReflectionImplementEmpty.h>
#include <NsCore/String.h>
#include <NsCore/TypeOf.h>
#include <NsCore/Boxing.h>

#include <math.h>


using namespace Menu3D;
using namespace Noesis;


////////////////////////////////////////////////////////////////////////////////////////////////////
bool MultiplierConverter::TryConvert(BaseComponent* value, const Type* targetType,
    BaseComponent* parameter, Ptr<BaseComponent>& result)
{
    if (targetType == TypeOf<float>() && value != 0 && parameter != 0)
    {
        float input = (float)atof(value->ToString().Str());
        float multiplier = (float)atof(parameter->ToString().Str());
        result.Reset(Boxing::Box(input * multiplier));
        return true;
    }

    return false;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
NS_BEGIN_COLD_REGION

NS_IMPLEMENT_REFLECTION_(Menu3D::MultiplierConverter, "Menu3D.MultiplierConverter")

NS_END_COLD_REGION

File: Samples/Menu3D/UE4/Content/MultiplierConverter.cs
using System;
using System.Globalization;
using System.Windows.Data;

namespace Menu3D
{
    public class MultiplierConverter : IValueConverter
    {
        public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
        {
            if (targetType == typeof(double) && value != null && parameter != null)
            {
                double input, multiplier;
                if (double.TryParse(value.ToString(), out input) &&
                    double.TryParse(parameter.ToString(), out multiplier))
                {
                    return input * multiplier;
                }
            }

            return value;
        }

        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
        {
            throw new NotImplementedException();
        }
    }
}
